internal/porchutil/function.go

```diff
@@ -0,0 +1,155 @@
+package porchutil
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/kubernetes"
+
+	// Enable the GCP Authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const expirationNanoSecond = 3
+
+var (
+	functionAPI     = "/apis/" + v1alpha1.SchemeGroupVersion.String() + "/functions"
+	PorchExpiration = time.Duration(expirationNanoSecond) * time.Second
+)
+
+// NewPorchFunctionClient provides the PorchFunctionClient which connects to Porch server
+// and gets the functions.
+func NewPorchFunctionClient() *PorchFunctionClient {
+	functions, err := visitFunctions()
+	if err != nil {
+		return &PorchFunctionClient{IsPorchAvailable: false, Functions: nil}
+	}
+	return &PorchFunctionClient{IsPorchAvailable: true, Functions: functions}
+}
+
+type PorchFunctionClient struct {
+	IsPorchAvailable bool
+	Functions        []v1alpha1.Function
+}
+
+// IsFunctionTypeMatch determines whether the `function` (which can be multi-typed), belongs
+// to the expectedType. type value should only be `validator` or `mutator`.
+func IsFunctionTypeMatch(function v1alpha1.Function, expectedType string) bool {
+	for _, actualType := range function.Spec.FunctionTypes {
+		if string(actualType) == expectedType {
+			return true
+		}
+	}
+	return false
+}
+
+// ContainsKeywords determines whether the `function` has keywords which match the `expectedKeywords`.
+// Experimental: This logic may change to only if all function keywords are found from  expectedKeywords,
+// can it claims a match (return true).
+func ContainsKeywords(function v1alpha1.Function, expectedKeywords []string) bool {
+	if len(expectedKeywords) == 0 {
+		// Accept all functions if keywords are not given.
+		return true
+	}
+	for _, actual := range function.Spec.Keywords {
+		for _, expected := range expectedKeywords {
+			if actual != expected {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ListKeywords returns all the unique keywords from Porch Functions.
+func (p *PorchFunctionClient) ListKeywords(cmd *cobra.Command) ([]string, error) {
```

Comment:
another way to do this could be that this function accepts:

`func match(*Function) bool`

and we can compose the matching on type and on keywords using that.

To me the 'red' flag is the interdependence of Porch Client (which conceptually can/should be purely around Porch API concepts) and cobra Commands, resulting in leaky abstraction

Reply:
really like this suggestion! thank you. Refactored the code.